Add gelu activation in pytorch

xiaomengy opened the pull request:
Summary:
Add gelu activation forward on CPU in pytorch

Compare to current python implemented version of gelu in BERT model like

  def gelu(self, x):
      x * 0.5 * (1.0 + torch.erf(x / self.sqrt_two))

The torch.gelu function can reduce the forward time from 333ms to 112ms (with MKL) / 133ms (without MKL) for input size = [64, 128, 56, 56] on a devvm.

Differential Revision: D15400974

soumith reviewed (changes requested):
looks pretty good!

You need to add documentation / stub function in functional.py
For example, see GLU : https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L954-L976

soumith commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1585,6 +1585,10 @@
     CUDA: prelu_backward_cuda
 
+- func: gelu(Tensor self) -> Tensor
+  dispatch:
```
you should add a `python_module: nn`. For example https://github.com/pytorch/pytorch/blob/6b8c1eab83100ae52c677f83786ea8dbbe2df5be/aten/src/ATen/native/native_functions.yaml#L3557-L3558

BIT-silence replied:
Thanks for pointing out this.

soumith commented on test/test_torch.py:
```diff
@@ -1769,6 +1770,23 @@ def matrixmultiply(mat1, mat2):
+                return X * norm.cdf(X)
+
+            X = np.random.randn(n, m).astype(dtype)
+            res = torch.gelu(torch.Tensor(X))
```
once you made the changes above, gelu will be `torch.nn.functional.gelu`

BIT-silence replied:
Done

fmassa commented:
Could you also add a GPU implementation for it? We try to keep parity between CPU and CUDA as much as possible.

Also, what's the story around the JIT fusion for CPU? It's disabled by default in PyTorch, but the fuser is currently able to handle all the operations in here inside a single fusion group, and seems to bring some significant speedups for CPU (and the same also applies to CUDA)

```python
import torch

# should we enable it by default?
torch._C._jit_override_can_fuse_on_cpu(True)

def gelu(x):
    sqrt_two = 1.4142135623730951
    return x * 0.5 * (1.0 + torch.erf(x / sqrt_two))

@torch.jit.script
def gelu2(x):
    sqrt_two = 1.4142135623730951
    return x * 0.5 * (1.0 + torch.erf(x / sqrt_two))

x = torch.rand(64, 128, 56, 56)

# compile gelu2
gelu2(x)
```
and gives
```
%timeit gelu(x)
> 92.6 ms ± 647 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit gelu2(x)
> 18.5 ms ± 518 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

fmassa commented on aten/src/ATen/native/cpu/Activation.cpp:
```diff
@@ -1,31 +1,81 @@
+void GeluKernelImpl(const Tensor& X, Tensor* Y) {
+  if (at::hasMKL()) {
+    AT_DISPATCH_FLOATING_TYPES(X.scalar_type(), "GeluKernelImpl", [&]() {
+      GeluKernelMKL<scalar_t>(X, Y);
```
Don't we need to check that the tensors are contiguous before dispatching to the MKL-optimized codepath?

gchanan replied:
This needs a comment; I would expect something like:
1) describe if MKL supports non-contiguous inputs or not.
2a) if it doesn't, when is it worth it to make the tensor contiguous to do the op?  Does it ever or should I just check contiguity in the MKL pass?
2b) if it does, why don't I just pass in the tensor?

BIT-silence replied:
I will add the cuda impl in the next PR. Thanks for the advice.

fmassa commented on test/test_torch.py:
```diff
@@ -1769,6 +1770,23 @@ def matrixmultiply(mat1, mat2):
+            def _gelu_ref(X):
+                return X * norm.cdf(X)
+
+            X = np.random.randn(n, m).astype(dtype)
```
Can you also test non-contiguous tensors? Like
```
x = torch.rand(50, 50)[:, ::2]
```

gchanan replied:
It looks like this always coverts the input to be contiguous, although it probably doesn't need to since it's now using TensorIterator.

BIT-silence replied:
I added test for non-contiguous inputs.

gchanan commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -371,4 +371,13 @@ Tensor hardshrink_backward_cpu(const Tensor & grad, const Tensor & self, Scalar
 }
 
+Tensor gelu_cpu(const Tensor& self) {
+  const auto X = self.contiguous();
```
why?

soumith replied:
is this `.contiguous()` still needed after your latest changes?

BIT-silence replied:
I think so, since currently our approach is using for-loop to take advantage of autovectorization for the non-mkl path. Actually currently with or without MKL, the performance is pretty similar.

BIT-silence commented:
For the current version, the MKL path in the same machine need 109ms while the non-MKL path need 112ms, currently they are quite similar.

cpuhrsch commented:
Will this in its final form also include backward and GPU support?

BIT-silence commented:
> Will this in its final form also include backward and GPU support?

Yes, I will add both backward and also the GPU support soon after this PR.

fmassa commented on aten/src/ATen/native/cpu/Activation.cpp:
```diff
@@ -1,31 +1,103 @@
+void GeluKernelImpl(const Tensor& X, Tensor* Y) {
+  if (at::hasMKL()) {
+    AT_DISPATCH_FLOATING_TYPES(X.scalar_type(), "GeluKernelImpl", [&]() {
+      GeluKernelMKLImpl<scalar_t>(X, Y);
```
Would using `TensorIterator` remove the need of using `MKL` functions altogether, and allow for computing everything without having to convert the tensors to contiguous?

BIT-silence replied:
I guess it will be a little tricky since the reason we try to use MKL is to use the v?CdfNorm function in MKL. So if we want to use TensorIterator here, I guess we have to implement the related things for CdfNorm first?

soumith replied:
that's not conflicting, as in you can use TensorIterator, where the specialization for a contiguous block can call v?CdfNorm if MKL is available. It's decoupled from using vec256.h.

But just do that in a follow-up diff, instead of this one, as it's an independent work unit anyways.

fmassa replied:
With `TensorIterator`, you could just bypass the need of `v?CdfNorm`, and use instead the methods that we have on `Vec256`, as we have `erf` implemented in `Vec256` as well.

So this would look something like, the following, which is copied from https://github.com/pytorch/pytorch/blob/052bab70697bf6cfe18570374ff26fce93dace94/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L33-L46

```c
static void gelu_kernel(TensorIterator& iter) {
  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "sigmoid_cpu", [&]() {
    unary_kernel_vec(
        iter,
        [=](scalar_t a) -> scalar_t {
          return (std::erf(a * (scalar_t)M_SQRT1_2) + (scalar_t)1) * a * (scalar_t)0.5; },
        [=](Vec256<scalar_t> a) {
          auto b = (Vec256<scalar_t>((scalar_t)(M_SQRT1_e)) * a;
          b = b.erf() + Vec256<scalar_t>((scalar_t)(1));
          return Vec256<scalar_t>((scalar_t)(0.5)) * a * b;
        });
  });
}
```

I'm not sure how much slower it would be compared to using `CdfNorm`, but this implementation seems simpler, handles non-contiguous tensors and uses vectorization thanks to sleef implementation of `erf`.

Or, as @soumith pointed out, one can also use `v?CdfNorm` on contiguous blocks, maybe taking inspiration from https://github.com/pytorch/pytorch/blob/052bab70697bf6cfe18570374ff26fce93dace94/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L190-L221

BIT-silence replied:
Actually I have tried to use TensorIterator with Vec256. In that case, it was still about 2x slower than the current approach on contiguous case.

BIT-silence replied:
I'm sorry that I made a mistake. Actually the reason that current approach using erf is faster is Tensor::erf() may also use MKL functions. If I disable MKL, the results are the same. So the conclusion is, if we have MKL, both using CdfNorm or Erf, we can be 2x faster than using Vec256 with MKL. I published https://github.com/pytorch/pytorch/pull/21326 to use TensorIterator. Please take a look. Thanks for the suggestions.

soumith commented on torch/nn/functional.py:
```diff
@@ -1148,6 +1148,19 @@ def rrelu(input, lower=1. / 8, upper=1. / 3, training=False, inplace=False):
+    Applies element-wise the function
+    :math:`\text{GeLU}(x) = x * \Phi(x)`
+
+    where `\Phi(x)` Cumulative Distribution Function for Gaussian Distribution.
```
where `\Phi(x)` is the Cumulative Distribution Function for Gaussian Distribution.

BIT-silence replied:
Done.

soumith commented on torch/nn/functional.py:
```diff
@@ -1148,6 +1148,19 @@ def rrelu(input, lower=1. / 8, upper=1. / 3, training=False, inplace=False):
+
+    where `\Phi(x)` Cumulative Distribution Function for Gaussian Distribution.
+
+    See :`Gaussian Error Linear Units (GELUs) <https://arxiv.org/abs/1606.08415>`.
```
add an entry in https://raw.githubusercontent.com/pytorch/pytorch/master/docs/source/nn.rst similar to elu:

```
:hidden:`gelu`
~~~~~~~~~~~~~

.. autofunction:: gelu
```

BIT-silence replied:
Thanks! Done.

soumith reviewed (approved):
ship when tests pass.

xiaomengy committed:
Add gelu activation in pytorch (#20665)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20665

Add gelu activation forward on CPU in pytorch

Compare to current python implemented version of gelu in BERT model like

  def gelu(self, x):
      x * 0.5 * (1.0 + torch.erf(x / self.sqrt_two))

The torch.nn.functional.gelu function can reduce the forward time from 333ms to 109ms (with MKL) / 112ms (without MKL) for input size = [64, 128, 56, 56] on a devvm.

Reviewed By: zheng-xq

Differential Revision: D15400974

fbshipit-source-id: 78399123aef803376a2459d487d44557126070ac

gchanan commented on aten/src/ATen/native/cuda/Activation.cu:
```diff
@@ -291,6 +293,24 @@ static void threshold_kernel(TensorIterator& iter, Scalar threshold, Scalar valu
+namespace {
+
+template <typename T>
+void GeluCUDAKernelImplInternal(const Tensor& X, Tensor* Y) {
```
can you avoid passing Tensors as `Tensor *`?  It's not standard C++ PyTorch code (is there any other example in the codebase that does this?).  Depending on the use case, you can use `const Tensor &`, `Tensor &` or `Tensor`.

BIT-silence replied:
I just want to make things clear that pass by pointer means it is a output variable or it will be changed in the function. Just for readability reason. I can change it to Tensor&.

https://google.github.io/styleguide/cppguide.html#Output_Parameters

gchanan replied:
See: https://github.com/pytorch/pytorch/wiki/Writing-Python-in-cpp-(a-manifesto)

Also, if you look at any `_out` function (translates to python with `out=` parameter), we use `Tensor&` already.  Although note that this kind of bogus, because reassigning the reference is almost never correct, but you should just follow the convention for now.

But the right way to think about this is you are already passing a (smart) pointer.  Passing a pointer to a smart pointer is almost never what you want.  And as noted in the link, const is essentially meaningless here, so trying to use static types for readability doesn't really work either (unless you implement ConstTensor).

BramVanroy commented:
Is gelu implemented in 1.2.0? I can find _some_ reference to it in the current documentation, but not in the API.

BramVanroy commented:
Is this implemented in 1.2.0? I can find it in documentation (https://pytorch.org/docs/stable/nn.functional.html) but I can't import it or find it in my installed library.

cpuhrsch commented:
@BramVanroy - is [this](https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.gelu) it?

BramVanroy commented:
> @BramVanroy - is [this](https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.gelu) it?

Odd. My IDE (PyCharm) underlines gelu in red and says "Cannot find reference 'gelu' in functional.pyi", but when I run the code it seems to import just fine.

```python
from torch.nn.functional import gelu
```

BramVanroy commented:
PyTorch's current implementation is

def gelu(x):
    x * 0.5 * (1.0 + torch.erf(x / self.sqrt_two))

rather than 

´´´python
def gelu(x):
    return 0.5 * x * (1 + torch.tanh(math.sqrt(math.pi / 2) * (x + 0.044715 * x ** 3)))
```

Correct?

BIT-silence commented:
PyTorch implementation is the original definition of GELU which is x * P(X <= x) where X ~ N(0, 1). This one is mathematically equivalent to 0.5 * x * (1.0 + torch.erf(x / math.sqrt(2.0))). The tanh is an approximation of GELU which may lead to better performance as mentioned in https://arxiv.org/pdf/1606.08415.pdf. However, in our test, the performance improvement depends on how tanh function is implemented. So we didn't use that by default.

BramVanroy commented:
@BIT-silence Thank you for the reply. That clarifies things a lot. A final question, is there a reason that gelu doesn't have a Module equivalent that lives in `nn`? (Like `nn.ReLU`.)

BIT-silence commented:
Actually the main reason is we don't have enough time when adding it. We will consider add the Module later.

BramVanroy commented:
Okay, thanks for the information. I wasn't sure whether in general there is a reason when an activation function has a `Module` equivalent, but it seems that it's mostly a time-constraint.

soumith commented:
@BramVanroy generally we only keep functionals for layers which don't have learnable parameters. We used to add layers for all common functions, like `nn.ReLU` but that's legacy

BramVanroy commented:
> @BramVanroy generally we only keep functionals for layers which don't have learnable parameters. We used to add layers for all common functions, like `nn.ReLU` but that's legacy

That makes sense. However, I do like that printing a Module, gives you a nice overview of its submodules. If the activation is a Module it is included (and clear to the user what it is). If it's a function, it's not included unfortunately - so you're left wondering whether (and where) an activation is taking place. Compare the two snippets below

```python
# as Module
from torch import nn

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.dense = nn.Linear(512, 1)
        self.activation = nn.ReLU()

    def forward(self, inputs):
        out = self.dense(inputs)
        out = self.activation(out)
        return out


net = Net()
print(net)
```

Prints

```
Net(
  (dense): Linear(in_features=512, out_features=1, bias=True)
  (activation): ReLU()
)
```

But when using functions, you don't see which activations are used nor where.


```python
from torch import nn
import torch.nn.functional as F


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.dense = nn.Linear(512, 1)
        self.activation = F.relu

    def forward(self, inputs):
        out = self.dense(inputs)
        out = self.activation(out)
        return out


net = Net()
print(net)
```

Prints

```
Net(
  (dense): Linear(in_features=512, out_features=1, bias=True)
)
```

BIT-silence commented:
The module which is torch.nn.GELU is implemented in https://github.com/pytorch/pytorch/pull/28944

BramVanroy commented:
Great! It'll ship with 1.4 then, I presume?
